fpt-egui/src/main.rs

```diff
@@ -256,218 +271,283 @@ impl FPT {
+                    $ui.colored_label(Color32::LIGHT_BLUE, $low_label);
+                }
+            }
+            let cpu = self.gb.cpu();
+            ui.vertical(|ui| {
+                Grid::new("cpu_registers_a-e").num_columns(4).min_col_width(10.0).striped(true).show(ui, |ui| {
+                    ui.colored_label(Color32::LIGHT_BLUE, "A");
+                    ui.monospace(format!("{:08b}", cpu.a()));
+                    ui.code(format!("{:#04X}", cpu.a()));
+                    ui.end_row();
+                    cpu_register!(ui, "B": cpu.b(), "C": cpu.c()); ui.end_row();
+                    cpu_register!(ui, "D": cpu.d(), "E": cpu.e()); ui.end_row();
+                    cpu_register!(ui, "H": cpu.a(), "L": cpu.f()); ui.end_row();
+                });
+            });
+            ui.separator();
+            ui.vertical(|ui| {
+                Grid::new("flags").num_columns(1).min_col_width(10.0).striped(true).show(ui, |ui| {
+                    ui.colored_label(Color32::LIGHT_BLUE, "Z");
+                    ui.code(if cpu.z_flag() { "1" } else { "0" });
+                    ui.colored_label(Color32::LIGHT_BLUE, "N");
+                    ui.code(if cpu.n_flag() { "1" } else { "0" });
+                    ui.end_row();
+                    ui.colored_label(Color32::LIGHT_BLUE, "H");
+                    ui.code(if cpu.h_flag() { "1" } else { "0" });
+                    ui.colored_label(Color32::LIGHT_BLUE, "C");
+                    ui.code(if cpu.c_flag() { "1" } else { "0" });
+                });
+                ui.horizontal(|ui| {
+                    ui.colored_label(Color32::LIGHT_BLUE, "SP");
+                    ui.code(format!("{:#06X}", cpu.sp()));
+                });
+                ui.horizontal(|ui| {
+                    ui.colored_label(Color32::LIGHT_BLUE, "PC");
+                    ui.code(format!("{:#06X}", cpu.pc()));
+                });
+            });
+        });
+        // TODO: scroll into line of current pc (need to find index)
+        ui.collapsing("Code", |ui| {
+            ScrollArea::vertical().show_rows(
+                ui,
+                ui.text_style_height(&egui::TextStyle::Body),
+                self.code.len(),
+                |ui, row_range| {
+                    for row in row_range {
+                        ui.label(RichText::new(self.code[row].1.clone()).monospace());
                     }
-                    for b in 0..TV_NUM_VBORDERS {
-                        for x in 0..TV_BORDER_SIZE {
-                            for y in 0..TV_Y_SIZE {
-                                self.tiles[(x + b * (TILE_SIZE + TV_BORDER_SIZE), y)] = GREY;
-                            }
+                },
+            );
+        });
+        ui.collapsing("Console", |ui| {
+            ScrollArea::vertical()
+                .auto_shrink(false)
+                .stick_to_bottom(true)
+                // TODO: dirty hack to make the console input always stick to the bottom
```

Comment:
This doesn't look like a big hack, it seems kinda natural, but I'll study egui's layouting mechanisms further